source/opt/combine_access_chains.cpp

```diff
@@ -0,0 +1,253 @@
+    new_value_id = addition->result_id();
+  } else {
+    // Indexing into structs must be constant, so bail out here.
+    return false;
+  }
+  new_operands->push_back({SPV_OPERAND_TYPE_ID, {new_value_id}});
+  return true;
+}
+
+bool CombineAccessChains::CreateNewOperands(
+    Instruction* ptr_input, Instruction* inst,
+    std::vector<Operand>* new_operands) {
+  // Start by copying all the input operands of the feeder access chain.
+  for (uint32_t i = 0; i != ptr_input->NumInOperands() - 1; ++i) {
+    new_operands->push_back(ptr_input->GetInOperand(i));
+  }
+
+  // Deal with the last index of the feeder access chain.
+  if (inst->opcode() == SpvOpPtrAccessChain) {
+    // The last index of the feeder should be combined with the element operand
+    // of |inst|.
+    if (!CombineIndices(ptr_input, inst, new_operands)) return false;
+  } else {
+    // The indices aren't being combined so now add the last index operand of
+    // |ptr_input|.
+    new_operands->push_back(
+        ptr_input->GetInOperand(ptr_input->NumInOperands() - 1));
+  }
+
+  // Copy the remaining index operands.
+  uint32_t starting_index = (inst->opcode() == SpvOpPtrAccessChain) ? 2 : 1;
+  for (uint32_t i = starting_index; i < inst->NumInOperands(); ++i) {
+    new_operands->push_back(inst->GetInOperand(i));
+  }
+
+  return true;
+}
+
+bool CombineAccessChains::CombineAccessChain(Instruction* inst) {
+  assert((inst->opcode() == SpvOpPtrAccessChain ||
+          inst->opcode() == SpvOpAccessChain ||
+          inst->opcode() == SpvOpInBoundsAccessChain) &&
+         "Wrong opcode. Expected an access chain.");
+
+  Instruction* ptr_input =
+      context()->get_def_use_mgr()->GetDef(inst->GetSingleWordInOperand(0));
+  if (ptr_input->opcode() != SpvOpAccessChain &&
+      ptr_input->opcode() != SpvOpInBoundsAccessChain &&
+      ptr_input->opcode() != SpvOpPtrAccessChain) {
+    return false;
+  }
+
+  // Handles the following cases:
+  // 1. |inst| is not a pointer access chain.
+  //    |inst|'s indices are appended to |ptr_input|'s indices.
+  // 2. |ptr_input| is not pointer access chain.
+  //    |inst|'s element operand is combined with the last index in
+  //    |ptr_input| to form a new operand.
+  // 3. |ptr_input| is a pointer access chain.
+  //    Like the above scenario, |inst|'s element operand is combined
+  //    with |ptr_input|'s last index. This results is either a
+  //    combined element operand or combined regular index.
+
+  // TODO(alan-baker): Support this properly. Requires analyzing the
+  // size/alignment of the type and converting the stride into an element
+  // index.
+  uint32_t array_stride = GetArrayStride(ptr_input);
+  if (array_stride != 0) return false;
+
+  std::vector<Operand> new_operands;
+  if (!CreateNewOperands(ptr_input, inst, &new_operands)) return false;
+
+  // Update the instruction.
+  if (ptr_input->opcode() == SpvOpInBoundsAccessChain &&
```

Comment:
This looks asymmetric enough to break down when OpInBoundsPtrAccessChain is supported.

Reply:
I re-factored this to use a helper.